### ISSUE TITLE
finetuning lears nothing

### ISSUE THREAD

**Issue by winstonquock** [deeplearning4j/deeplearning4j#203]
1) with conjugate gradient descent, there are two problems: the EpsTerminationCondition would stop the iterations due to little differences; removing the eps termination condition, the scores are not changing between the iterations.
```
// conjugate gradient descent, no eps termination condition
output weight before fit
[
  [1.5101332974154502E-4, 0.009033489041030407, -0.009212066419422626, 1.4481108519248664E-4, 0.003211055416613817]
  [0.010129489004611969, -0.0024550841189920902, 0.010258954018354416, -0.015858318656682968, 0.00983279850333929]
  [0.00610418850556016, -0.01347651518881321, 0.010372709482908249, -0.001226464519277215, -0.007650960236787796]
  [-0.011538776569068432, -0.009486142545938492, -0.011927636340260506, 2.1848703909199685E-4, 0.0023736569564789534]
  [0.0013413570122793317, 0.0048651681281626225, 0.011178816668689251, -0.005712731741368771, -0.003965042065829039]
] // 5x5
o.d.o.l.ScoreIterationListener - Score at iteration 0 is 0.40000002617662167
o.d.o.l.ScoreIterationListener - Score at iteration 1 is 0.40000002617662167
o.d.o.l.ScoreIterationListener - Score at iteration 2 is 0.40000002617662167
o.d.o.l.ScoreIterationListener - Score at iteration 3 is 0.40000002617662167
o.d.o.l.ScoreIterationListener - Score at iteration 4 is 0.40000002617662167

output weight after fit
[
  [1.5101332974154502E-4, 0.009033489041030407, -0.009212066419422626, 1.4481108519248664E-4, 0.003211055416613817]
  [0.010129489004611969, -0.0024550841189920902, 0.010258954018354416, -0.015858318656682968, 0.00983279850333929]
  [0.00610418850556016, -0.01347651518881321, 0.010372709482908249, -0.001226464519277215, -0.007650960236787796]
  [-0.011538776569068432, -0.009486142545938492, -0.011927636340260506, 2.1848703909199685E-4, 0.0023736569564789534]
  [0.0013413570122793317, 0.0048651681281626225, 0.011178816668689251, -0.005712731741368771, -0.003965042065829039]
] // 5x5
```

2) with iterative gradient descent, uncomment the model update (org.deeplearning4j.optimize.solvers.IterationGradientDescent, line 50), the score are changing but the network ends up predicting only one class for all inputs
```
output weight before fit
[
  [1.5101332974154502E-4, 0.009033489041030407, -0.009212066419422626, 1.4481108519248664E-4, 0.003211055416613817]
  [0.010129489004611969, -0.0024550841189920902, 0.010258954018354416, -0.015858318656682968, 0.00983279850333929]
  [0.00610418850556016, -0.01347651518881321, 0.010372709482908249, -0.001226464519277215, -0.007650960236787796]
  [-0.011538776569068432, -0.009486142545938492, -0.011927636340260506, 2.1848703909199685E-4, 0.0023736569564789534]
  [0.0013413570122793317, 0.0048651681281626225, 0.011178816668689251, -0.005712731741368771, -0.003965042065829039]
] // 5x5
o.d.o.l.ScoreIterationListener - Score at iteration 0 is 0.39999863009013137
o.d.o.s.BaseOptimizer - Error at iteration 0 was 0.39999863009013137
o.d.o.l.ScoreIterationListener - Score at iteration 1 is 0.39140733196703004
o.d.o.s.BaseOptimizer - Error at iteration 1 was 0.39140733196703004
o.d.o.l.ScoreIterationListener - Score at iteration 2 is 0.38592743072618835
o.d.o.s.BaseOptimizer - Error at iteration 2 was 0.38592743072618835
o.d.o.l.ScoreIterationListener - Score at iteration 3 is 0.38249092020248837
o.d.o.s.BaseOptimizer - Error at iteration 3 was 0.38249092020248837
o.d.o.l.ScoreIterationListener - Score at iteration 4 is 0.3803493235262911
o.d.o.s.BaseOptimizer - Error at iteration 4 was 0.3803493235262911
output weight after fit
[
  [-2.8183116228319705E-4, 0.008882595226168633, -0.009172262623906136, 5.196834681555629E-4, 0.0033801181707531214]
  [0.010077212937176228, -0.002507733879610896, 0.01037511508911848, -0.015912357717752457, 0.009875602088868618]
  [0.006043777335435152, -0.013579457066953182, 0.010519091039896011, -9.845653548836708E-4, -0.00787588581442833]
  [-0.011889117769896984, -0.00961777288466692, -0.01180428545922041, 6.061353487893939E-4, 0.0023446313571184874]
  [0.0011195194674655795, 0.004625934176146984, 0.011281772516667843, -0.005316227208822966, -0.004003431182354689]
] // 5x5

training set labels
[
  [1.0, 0.0, 0.0, 0.0, 0.0]
  [0.0, 0.0, 0.0, 0.0, 1.0]
  [0.0, 1.0, 0.0, 0.0, 0.0]
  [1.0, 0.0, 0.0, 0.0, 0.0]

  ...
  [0.0, 0.0, 1.0, 0.0, 0.0]
  [0.0, 0.0, 0.0, 1.0, 0.0]
  [0.0, 0.0, 0.0, 0.0, 1.0]
  [0.0, 0.0, 0.0, 1.0, 0.0]
  [0.0, 1.0, 0.0, 0.0, 0.0]
] // 1399x5
predicted (against the same training set)
[
  [0.1937110871076584, 0.39966434240341187, -0.3920721411705017, -0.13856546580791473, -0.06563594192266464]
  [0.1937110871076584, 0.39966434240341187, -0.3920721411705017, -0.13856545090675354, -0.06563594192266464]
  [0.1937110871076584, 0.3996642231941223, -0.39207205176353455, -0.13856549561023712, -0.06563599407672882]
  [0.1937110871076584, 0.3996642827987671, -0.3920721113681793, -0.13856539130210876, -0.06563600152730942]
  ...
  [0.1937110275030136, 0.3996642231941223, -0.3920721113681793, -0.13856551051139832, -0.06563598662614822]
  [0.1937110871076584, 0.3996642231941223, -0.39207205176353455, -0.1385655552148819, -0.06563598662614822]
  [0.1937110722064972, 0.3996642827987671, -0.3920721113681793, -0.1385655403137207, -0.06563592702150345]
  [0.1937110722064972, 0.3996642827987671, -0.3920721113681793, -0.1385655552148819, -0.06563592702150345]
  [0.1937110722064972, 0.3996641933917999, -0.3920721113681793, -0.1385655403137207, -0.06563597917556763]
] // 1399x5

F1: 0.10297239915074309 precision: 0.06933523945675482. recall: 0.2. accuracy: 0.5685626622610338
  CLASS[0]: precision: 0.0 recall: 0.0
  CLASS[1]: precision: 0.34667619728377413 recall: 1.0
  CLASS[2]: precision: 0.0 recall: 0.0
  CLASS[3]: precision: 0.0 recall: 0.0
  CLASS[4]: precision: 0.0 recall: 0.0
```

**Comment by agibsonccc** [comment]
I fixed it for Iteration Gradient Descent. I think in one of the updates (yours?) I pulled in the commented update. For whatever reason you commented out the update. I'm taking a second look at conjugate gradient, but I don't necessarily think there's a problem there.

**Comment by winstonquock** [comment]
Yeah, I made the change to the IGD but I indicated it in the PR about this potential problem. Will try the latest changes

**Comment by agibsonccc** [comment]
Cool thanks. Yeah things happen. Either way it got sorted out. Appreciate the help!

**Comment by winstonquock** [comment]
Please re-open. With the latest codes, the same problems still plague the CGD and IGD

1) CGD: eps termination still terminate at the first iteration. The weigth gradient still does not change.

```
before finetune, output W
[
  [-0.008826193766912635, 0.004481822061215228, -0.004978411060519378, -0.0033180523774239052, 0.013983840475942097]
  [-3.127776605890697E-4, 7.157517392885522E-4, 0.002499666748289063, 4.0898384483290586E-4, 0.0020989381456352815]
  [0.0024380961485814725, 0.009340209620062712, 0.00534521028378077, 2.5232994024476292E-5, 0.0031098789256406474]
  [-0.00660829774499396, -1.0892056590344007E-4, -0.006346663972570686, 0.010388207040575453, 0.00284442042526733]
  [-0.004885078679115029, 0.007566454511438661, 0.01266812077317913, 0.007409378409097654, 0.010530146471866763]
] // 5x5
o.d.o.l.ScoreIterationListener - Score at iteration 0 is 0.40000032865020607
 after finetune, output W
[
  [-0.008826193766912635, 0.004481822061215228, -0.004978411060519378, -0.0033180523774239052, 0.013983840475942097]
  [-3.127776605890697E-4, 7.157517392885522E-4, 0.002499666748289063, 4.0898384483290586E-4, 0.0020989381456352815]
  [0.0024380961485814725, 0.009340209620062712, 0.00534521028378077, 2.5232994024476292E-5, 0.0031098789256406474]
  [-0.00660829774499396, -1.0892056590344007E-4, -0.006346663972570686, 0.010388207040575453, 0.00284442042526733]
  [-0.004885078679115029, 0.007566454511438661, 0.01266812077317913, 0.007409378409097654, 0.010530146471866763]
] // 5x5
```

2) for IGD, the network now predicts only 2 classes (instead of 1 previously.)
```
before finetune, output W
[
  [0.008316120224736257, 0.0019454015149312437, -0.011335602216057752, 0.0027272115545495836, -0.005706872122389523]
  [-0.002512422786772838, -0.0037372909603617583, 0.0016370028025516066, 0.003075214797774339, -0.0014851003161337473]
  [-0.0015627904685785387, -0.0015201375087366598, 0.004742828832334199, -0.0028970179249528604, 0.004123426038942327]
  [-0.003107336047333228, -0.001623997952740581, 0.008046103826546816, 0.005226728767668746, -0.0021585440851123613]
  [-0.012805317704534072, -0.011641106903908374, 0.00912513610592341, -0.01591781554664934, 0.019460671719353062]
] // 5x5
o.d.o.l.ScoreIterationListener - Score at iteration 0 is 0.40082180118625277
o.d.o.l.ScoreIterationListener - Score at iteration 1 is 0.7862759113652609
o.d.o.l.ScoreIterationListener - Score at iteration 2 is 0.7862759113652609
o.d.o.l.ScoreIterationListener - Score at iteration 3 is 0.7862759113652609
o.d.o.l.ScoreIterationListener - Score at iteration 4 is 0.7862759113652609
 after finetune, output W
[
  [-112.96085764249081, -2201.363943587116, -4.146031785773104, 2410.1425098625764, -91.67573058823947]
  [-112.9716861855023, -2201.369626279591, -4.133059180754493, 2410.1428578658197, -91.67150881643322]
  [-112.97073655318412, -2201.3674091261396, -4.129953354724712, 2410.136885633097, -91.66590029007814]
  [-112.97228109876288, -2201.3675129865833, -4.126650079730499, 2410.1450093797894, -91.67218226020219]
  [-112.98197908042007, -2201.377530095535, -4.125571047451121, 2410.123864835475, -91.65056304439773]
] // 5x5

F1: 0.1133229770798573 precision: 0.0809492380940521. recall: 0.18884846932240298. accuracy: 0.37281067556296915
  CLASS[0]: precision: 0.2724306688417618 recall: 0.45135135135135135
  CLASS[1]: precision: 0.0 recall: 0.0
  CLASS[2]: precision: 0.0 recall: 0.0
  CLASS[3]: precision: 0.13231552162849872 recall: 0.4928909952606635
  CLASS[4]: precision: 0.0 recall: 0.0

```

**Comment by agibsonccc** [comment]
What dataset is this? I'm going to need some example code, I tried it on iris and it's been fine. I'm not sure if this has anything to do with a bug. Some things just don't have signal.

Also: not to call you out here, but you removing the eps termination on the changes makes no sense. Of course the weights don't change, that means its converged.
